crates/ruff_linter/src/checkers/noqa.rs

```diff
@@ -51,8 +51,15 @@ pub(crate) fn check_noqa(
 
         match &exemption {
             FileExemption::All(_) => {
-                // If the file is exempted, ignore all diagnostics.
-                ignored_diagnostics.push(index);
+                // If the file is exempted, ignore all diagnostics,
+                // save for RUF104, which operates on `# noqa` comments
+                // and thus needs to be suppressed explicitly.
+                if !matches!(diagnostic.kind.rule(), Rule::UnformattedSpecialComment)
```

Comment:
Can you explain the reasoning for this special handling? I think I would find this behavior surprising and I'm not sure if it justifies the added complexity